src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,389 @@
+	up: 38,
+	down: 40
+};
+
+export const ComboBoxBase = ThemeableMixin(WidgetBase);
+
+@theme(css)
+export default class ComboBox extends ComboBoxBase<ComboBoxProperties> {
+	private _ignoreBlur: boolean;
+	private _ignoreFocus: boolean;
+	private _inputElement: HTMLInputElement;
+	private _focused: boolean;
+	private _open: boolean;
+	private _direction: Operation;
+	private _selectedIndex: number | undefined;
+	private _wasOpen: boolean;
+
+	private _handlers: {[key: string]: any} = {
+		[keys.up](this: ComboBox, event: KeyboardEvent) {
+			event.preventDefault();
+			this.updateSelectedIndex(Operation.decrease);
+			this.invalidate();
+		},
+
+		[keys.down](this: ComboBox, event: KeyboardEvent) {
+			event.preventDefault();
+			this.updateSelectedIndex(Operation.increase);
+			this.invalidate();
+		},
+
+		[keys.enter](this: ComboBox) {
+			const { results } = this.properties;
+
+			// If results already closed or no results, do nothing
+			if (!this._open || !results || results.length === 0) {
+				return;
+			}
+			// If no result selected, close menu
+			else if (this._selectedIndex === undefined) {
+				this._open = false;
+				this.invalidate();
+			}
+			else {
+				this.selectResult(results[this._selectedIndex]);
+			}
+		},
+
+		[keys.escape](this: ComboBox) {
+			this.updateSelectedIndex(Operation.reset);
+			this._open = false;
+			this.invalidate();
+		}
+	};
+
+	updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const increment = operation === Operation.increase ? 1 : -1;
+				const total = results.length;
+				const current = this._selectedIndex;
+				const base = operation === Operation.increase ? -1 : 0;
+
+				this._selectedIndex = ((current === undefined ? base : current) + increment + total) % total;
+			}
+			else {
+				this._selectedIndex = undefined;
+			}
+			this._direction = operation;
+		}
+	}
+
+	createRegistry() {
+		const {
+			customResultItem = ResultItem,
+			customResultMenu = ResultMenu
+		} = this.properties;
+
+		const registry = new FactoryRegistry();
+		registry.define('result-item', customResultItem);
+		registry.define('result-menu', customResultMenu);
+		return registry;
+	}
+
+	@onPropertiesChanged
+	onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
+		if (!this.registry || includes(evt.changedPropertyKeys, 'customResultItem') || includes(evt.changedPropertyKeys, 'customResultMenu')) {
+			this.registry = this.createRegistry();
+		}
+	}
+
+	afterCreate(element: HTMLElement) {
+		this._inputElement = <HTMLInputElement> element.querySelector('input');
+		// Maintain focused state
+		this._inputElement && this._inputElement[this._focused ? 'focus' : 'blur']();
```

Comment:
Do you think it would be clearer to extract ternary out from the inline statement? Or perhaps add a helper function as it looks like it might be used a few times?